### ISSUE TITLE
Proposal: ODIC/Oauth2 client for Remote Write.

### ISSUE THREAD

**Issue by bwplotka** [prometheus/prometheus#8132]
Currently, it's extremely common to have an OIDC/Oauth2 based server auth, especially for SaaS offerings. 

This is why I would propose to add OIDC/Oauth2 client to Prometheus remote write endpoint. Those protocols are well known and well used, so I would expect high usage of those. 

I searched through docs and couldn't find the related discussion so starting one here.

I know you can leverage Prometheus by replacing the bearer token in the configuration of remote write (https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write) but this request is to add native support.

By any chance, is there any sidecar that already integrates with Prometheus through config or as a front proxy?

**Comment by bwplotka** [comment]
Mhm, I can answer myself, it's a no for current time being: https://www.robustperception.io/prometheus-security-authentication-authorization-and-encryption so moving to mailing list to look for potential sidecars :hugs:

**Comment by bwplotka** [comment]
https://groups.google.com/g/prometheus-users/c/LoVWVwNhu6o

**Comment by brian-brazil** [comment]
From previously looking into these for other stuff, my understanding of these providers is that you need to implement code for each one specifically (i.e. one for Google, one for Facebook etc.) and these flows also require human interaction for initial setup. I don't think that Oauth2 is compatible with Prometheus's architecture or deployment model, you'd need to have something else provide the token to Prometheus.

**Comment by roidelapluie** [comment]
Yes that is possible, but you have that already. To have this working, you need a oauth token, as Prometheus is static, you would generate a service token in Keycloak, which does not change and can use the existing Token or Token file options we have.

**Comment by roidelapluie** [comment]
https://wjw465150.gitbooks.io/keycloak-documentation/content/server_admin/topics/clients/oidc/service-accounts.html

It turns out that we would still need to fetch the token.

**Comment by bwplotka** [comment]
The idea would be to just use OIDC client to do so like https://github.com/bwplotka/oidc

It works against any OIDC compatible backends

**Comment by csmarchbanks** [comment]
OIDC is common enough that I think it is worthwhile adding native support for service accounts. It is certainly possible as I have implemented OIDC in a remote-write adapter and fluentd plugins before.
